src/python/pants/engine/legacy/parser.py

```diff
@@ -125,5 +128,32 @@ def parse(self, filepath, filecontent):
     # _intentional_ mutation would require a deep clone, which doesn't seem worth the cost at
     # this juncture.
     self._parse_context._storage.clear(os.path.dirname(filepath))
-    six.exec_(python, dict(self._symbols))
+    symbols = dict(self._symbols)
+
+    options, _ = OptionsInitializer(OptionsBootstrapper()).setup()
+    global_options = options.for_global_scope()
+    if global_options.build_file_imports != 'allow':
+      # This is not secure sandboxing, because people could replace __import__ again themselves
+      # if they wanted (and there are plenty of other ways to escape a python "sandbox"), but it
+      # should be sufficient to tell the casual user that they're doing something wrong.
+      builtins = dict(__builtins__)
+      if global_options.build_file_imports == 'warn':
+        import_hook = lambda import_name, *args: _warn_on_import(__import__, filepath, import_name, *args)
+      elif global_options.build_file_imports == 'error':
+        import_hook = _fail_on_import
+      else:
+        raise Exception("Didn't know what to do for build_file_imports value {}".format(global_options.build_file_imports))
+      builtins['__import__'] = import_hook
+      symbols['__builtins__'] = builtins
+
+    six.exec_(python, symbols)
     return list(self._parse_context._storage.objects)
+
+
+def _warn_on_import(import_builtin, filepath, import_name, *args):
+  logging.getLogger(__name__).warn('BUILD file at {} tried to import {} - import statements should be avoided'.format(filepath, import_name))
+  return import_builtin(import_name, *args)
+
+
+def _fail_on_import(name, *args):
+  raise Exception('import statements have been banned, but tried to import {}'.format(name))
```

Comment:
Very probably... `pants.engine.parser.ParseError`, `pants.build_graph.build_file_parser.ParseError`, `pants.base.exceptions.TargetDefinitionException`, or a new subclass of `pants.base.build_file.BuildFileError` all seem like viable options.

I'm guessing `pants.engine.parser.ParseError` is probably the winner, because this file already uses things in that module? I've switched to that one for now...